packages/authorization-server/src/services/tokenService.ts

```diff
@@ -0,0 +1,442 @@
+  redisRateLimiter: RateLimiter;
+  producer: Awaited<ReturnType<typeof initProducer>>;
+  fileManager: FileManager;
+}) {
+  return {
+    async generateToken(
+      request: authorizationServerApi.AccessTokenRequest,
+      correlationId: string,
+      logger: Logger
+    ): Promise<GenerateTokenReturnType> {
+      const { errors: parametersErrors } = validateRequestParameters({
+        client_assertion: request.client_assertion,
+        client_assertion_type: request.client_assertion_type,
+        grant_type: request.grant_type,
+        client_id: request.client_id,
+      });
+
+      if (parametersErrors) {
+        throw clientAssertionRequestValidationFailed(request);
+      }
+
+      const { data: jwt, errors: clientAssertionErrors } =
+        verifyClientAssertion(request.client_assertion, request.client_id);
+
+      if (clientAssertionErrors) {
+        throw clientAssertionValidationFailed(
+          request.client_assertion,
+          request.client_id
+        );
+      }
+
+      const clientId = jwt.payload.sub;
+      const kid = jwt.header.kid;
+      const purposeId = jwt.payload.purposeId;
+
+      const pk = purposeId
+        ? makeTokenGenerationStatesClientKidPurposePK({
+            clientId,
+            kid,
+            purposeId,
+          })
+        : makeTokenGenerationStatesClientKidPK({ clientId, kid });
+
+      const key = await retrieveKey(dynamoDBClient, pk);
+
+      const { errors: clientAssertionSignatureErrors } =
+        await verifyClientAssertionSignature(
+          request.client_assertion,
+          key,
+          jwt.header.alg
+        );
+
+      if (clientAssertionSignatureErrors) {
+        throw clientAssertionSignatureValidationFailed(
+          request.client_assertion
+        );
+      }
+
+      const { errors: platformStateErrors } =
+        validateClientKindAndPlatformState(key, jwt);
+      if (platformStateErrors) {
+        throw platformStateValidationFailed();
+      }
+
+      const { limitReached, ...rateLimiterStatus } =
+        await redisRateLimiter.rateLimitByOrganization(key.consumerId, logger);
+      if (limitReached) {
+        return {
+          limitReached: true,
+          token: undefined,
+          rateLimitedTenantId: key.consumerId,
+          rateLimiterStatus,
+        };
+      }
+
+      return await match(key.clientKind)
+        .with(clientKindTokenStates.consumer, async () => {
+          if (key.PK.startsWith(clientKidPurposePrefix)) {
+            const parsedKey = key as TokenGenerationStatesClientPurposeEntry;
+
+            if (!parsedKey.descriptorAudience || !parsedKey.GSIPK_purposeId) {
+              throw invalidTokenClientKidPurposeEntry();
+            }
+            const token = await tokenGenerator.generateInteropConsumerToken({
+              sub: jwt.payload.sub,
+              audience: parsedKey.descriptorAudience,
+              purposeId: parsedKey.GSIPK_purposeId,
+              tokenDurationInSeconds: 10,
```

Comment:
`tokenDurationInSeconds` is be the `voucherLifespan` in the DynamoDB record

Reply:
Fixed